lib/src/room.dart

```diff
@@ -223,32 +221,57 @@ class Room {
     return pinned is Iterable ? pinned.map((e) => e.toString()).toList() : [];
   }
 
-  /// Returns a localized displayname for this server. If the room is a groupchat
+  /// Returns the heroes as `User` objects.
+  Future<List<User>> getHeroUsers({
+    bool filterOwnUser = true,
+    bool dmUserFallback = true,
+  }) async {
+    var heroes = summary.mHeroes;
+    if (heroes == null && dmUserFallback) {
+      final directChatMatrixID = this.directChatMatrixID;
+      if (directChatMatrixID != null) {
+        heroes = [directChatMatrixID];
+      }
+    }
+
+    if (heroes == null) return [];
+
+    if (filterOwnUser) heroes.remove(client.userID!);
+
+    return await Future.wait(heroes.map((hero) async =>
+        (await requestUser(
+          hero,
+          ignoreErrors: true,
+        )) ??
+        User(hero, room: this)));
+  }
+
+  /// Returns a localized displayname for this room. If the room is a groupchat
   /// without a name, then it will return the localized version of 'Group with Alice' instead
   /// of just 'Alice' to make it different to a direct chat.
   /// Empty chats will become the localized version of 'Empty Chat'.
   /// This method requires a localization class which implements [MatrixLocalizations]
-  String getLocalizedDisplayname([
+  Future<String> getLocalizedDisplayname([
```

Comment:
I would actually prefer to cache this (and member count, room avatar, encryption status as well as membership) in a room info object and only load that when loading the room, and load all events lazily when requested instead.

Reply:
Not sure if I understand this correctly. You mean that we have a `await room.getInfo()` where then the calculated displayname and avatar is inside?

Reply:
More like, when we load the room from the database, it already loads the cached room name and avatar, so you can return it immediately and will not need to load any member events unless you want to update the room name (i.e. on sync).

Reply:
This was the behavior with the old database but that was super slow because it did not scale that good. If we have 1000 rooms, then this will do 1000*x requests (x==heroes) and block the app start. While we actually only need the first 10 or so rooms as long as the user does not scroll down the chat list. Loading the missing users on demand has a much better performance

Reply:
Other suggestion: We leave the methods as they are but add a method `room.loadHeroUsers()` and `bool get heroUsersAreCached`. So in the UI we could react on this instead.

But that would make it possible to mess it up in the app because the sdk consumer would need to know about it...

Reply:
No, you wouldn't load the heroes at all, you would only load them on /sync if the room name needs to be recalculated (because one of the heroes member events changed or the room name was set/unset, etc).

Basically you would calculate the room name and avatar once and then store it in the toJson/fromJson calls: https://github.com/famedly/matrix-dart-sdk/blob/94313a09047c444f332b8ec85cb8374262f6f662/lib/src/room.dart#L95

The calculation happens whenever any of the state contributing to the room name changes, which usually happens mostly on initial sync. Then whenever we load the room from the database, we just load 2 additional string values instead of loading each member event, which shouldn't add much overhead, since we already load id, prev_batch, notification counts and summary.

Reply:
I've thought about this a lot and I'm not 100% sure if this fits our needs. There are advantages and disadvantages. However this would be a huge change to the sdk and the database and it might make more sense to do this when we have reduced the amount of databases which are in use.

What I currently would like to fix is the current behavior that we already lazy load the hero users but we do not know when to update the UI, because calling `room.getLocalizedDisplayname()` actually requests the user but we do not know when this is done to update the UI